File: common/src/main/kotlin/net/impleri/playerskills/utils/ListDiff.kt
package net.impleri.playerskills.utils

import net.minecraft.resources.ResourceLocation

object ListDiff {
  fun <T> contains(a: List<T>, b: List<T>, getName: (T) -> ResourceLocation): Boolean {
    return getMissing(a, b, getName).isEmpty()
  }

  fun <T> getMissing(a: List<T>, b: List<T>, getName: (T) -> ResourceLocation): List<T> {
    if (a.isEmpty() && b.isEmpty()) {
      return ArrayList()
    }

    val bStrings = b.map { getName(it) }

    return a.filter { !bStrings.contains(getName(it)) }
  }
}

File: common/src/main/scala/net/impleri/playerskills/utils/ListDiff.scala
package net.impleri.playerskills.utils

import net.minecraft.resources.ResourceLocation

object ListDiff {
  def contains[T](a: Seq[T], b: Seq[T], getName: T => ResourceLocation): Boolean = diff(a, b, getName).nonEmpty

  // TODO: Maybe Seq.diff is enough and this is not necessary?
  def diff[T](a: Seq[T], b: Seq[T], getName: T => ResourceLocation): Seq[T] = {
    val names = diffNames(a, b, getName)
    a.filter(v => names.contains(getName(v)))
  }

  private def diffNames[T](a: Seq[T], b: Seq[T], getName: T => ResourceLocation) = a.map(getName).diff(b.map(getName))
}
